gapidapk/deviceinfo.go

```diff
@@ -25,15 +28,117 @@ import (
-	sendDevInfoPort    = "gapid-devinfo"
+	sendDevInfoAction           = "com.google.android.gapid.action.SEND_DEV_INFO"
+	sendDevInfoService          = "com.google.android.gapid.DeviceInfoService"
+	sendDevInfoPort             = "gapid-devinfo"
+	startServiceAttempts        = 3
+	portListeningTimeoutSeconds = 5 * time.Second
 )
 
 func init() {
 	adb.RegisterDeviceInfoProvider(fetchDeviceInfo)
 }
 
+// Returns true if the device is listening to sendDevInfoPort, false if not.
+// Error if failed at getting the port info.
+func devInfoPortListening(ctx context.Context, d adb.Device) (bool, error) {
+	var stdout bytes.Buffer
+	if err := d.Shell("cat", "/proc/net/unix").Capture(&stdout, nil).Run(ctx); err != nil {
+		return false, log.Errf(ctx, err, "Getting unix abstract port info...")
+	}
+	if strings.Contains(stdout.String(), sendDevInfoPort) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// startDevInfoService tries to start the fresh run of the package and start
+// the service to send device info.
+func startDevInfoService(ctx context.Context, d adb.Device, apk *APK) error {
+	ctx = log.Enter(ctx, "startDevInfoService")
+	var se error
+	var le error
+	var listening bool
+
+	action := apk.ServiceActions.FindByName(sendDevInfoAction, sendDevInfoService)
+	if action == nil {
+		return log.Err(ctx, nil, "Service intent was not found")
+	}
+
+	// Try to start service.
+	for i := 0; i < startServiceAttempts; i++ {
+		log.I(ctx, "Attempt to start service: %s", sendDevInfoService)
+		if se = d.StartService(ctx, *action); se != nil {
+			continue
+		}
+
+		stop_checking_port := make(chan bool)
+		timer := time.AfterFunc(portListeningTimeoutSeconds, func() {
+			stop_checking_port <- true
+			close(stop_checking_port)
+		})
+		func() {
+			select {
+			case <-stop_checking_port:
+				return
+			default:
+				log.I(ctx, "Checking port: %s on device %s", sendDevInfoPort, d.Instance().GetSerial())
+				listening, le = devInfoPortListening(ctx, d)
+				if le != nil || listening {
+					return
+				}
+				time.Sleep(time.Second)
+			}
+		}()
+		timer.Stop()
+		if listening {
+			return nil
+		}
+	}
+	if se != nil {
+		return se
+	}
+	if le != nil {
+		return le
+	}
+	return log.Errf(ctx, nil, "Run out of attempts: %v", startServiceAttempts)
+}
+
+// Checks the existence of VkGraphicsSpyLayer in the debug.vulkan.layers system
+// property. If found, strip the layer, otherwise keep the property unchanged.
+// Returns a property value recover callback and error.
+func stripVkGraphicsSpyLayer(ctx context.Context, d adb.Device) (func(), error) {
+	const propName = "debug.vulkan.layers"
+	const layerName = "VkGraphicsSpy"
+	ctx = log.Enter(ctx, "stripVkGraphicsSpyLayer")
+	log.I(ctx, "Check the existence of %s in %s", layerName, propName)
+	old_layer_str, err := d.SystemProperty(ctx, propName)
+	if err != nil {
+		return nil, log.Errf(ctx, err, "Getting %s.", propName)
+	}
+	should_strip := false
+	old_layers := strings.Split(old_layer_str, ":")
```

Comment:
Don't use underscores in variable names. Please use golint and govet as part of your development workflow.